src/dodal/devices/current_amplifiers/current_amplifier_detector.py

```diff
@@ -0,0 +1,110 @@
+        auto_mode (signalR([bool])): Soft signal to store the flag for auto gain.
+        upper_limit (float): The upper limit of the current amplifier output in volt.
+        lower_limit (float): The lower limit of the current amplifier output in volt.
+        name (str): Name of the device.
+    """
+
+    def __init__(
+        self,
+        current_amp: CurrentAmp,
+        counter: CurrentAmpCounter,
+        upper_limit: float,
+        lower_limit: float,
+        name: str = "",
+    ) -> None:
+        self.current_amp = Reference(current_amp)
+        self.counter = Reference(counter)
+        with self.add_children_as_readables():
+            self.current, self._current_set = soft_signal_r_and_setter(
+                float, initial_value=None, units="Amp"
+            )
+        with self.add_children_as_readables(StandardReadableFormat.CONFIG_SIGNAL):
+            self.auto_mode = soft_signal_rw(bool, initial_value=True)
+            self.upper_limit = upper_limit
+            self.lower_limit = lower_limit
+        super().__init__(name)
+
+    async def read(self) -> dict[str, Reading]:
+        """
+        Read is modified so that if auto_mode is true it will optimism gain before
+         taking the final reading
+        """
+        if await self.auto_mode.get_value() is True:
+            LOGGER.info(f"{self.name}-Attempting auto-gain")
+            if await self.auto_gain():
+                LOGGER.info(
+                    f"{self.name} new gain = {await self.current_amp().get_gain()}."
+                )
+            else:
+                LOGGER.warning(f"{self.name} new gain is at maximum/minimum value.")
+        current = await self.get_corrected_current()
+        self._current_set(current)
+        return await super().read()
+
+    async def auto_gain(self) -> bool:
+        cnt = 0
+        while cnt < len(self.current_amp().gain_conversion_table):
+            """
+            negative value is possible on some current amplifier, hence the abs.
+            """
+            reading = abs(await self.counter().get_voltage_per_sec())
+            if reading > self.upper_limit:
+                if not await self.current_amp().decrease_gain():
+                    return False
+            elif reading < self.lower_limit:
+                if not await self.current_amp().increase_gain():
+                    return False
+            else:
+                break
+            cnt += 1
+        return True
+
+    async def get_corrected_current(self) -> float:
+        """
+        Convert the output(count and gain) back into the read detector output in Amp.
+        """
+        current_gain = await self.current_amp().get_gain()
+        correction_factor = self.current_amp().gain_conversion_table[current_gain].value
+        corrected_current = (
+            await self.counter().get_voltage_per_sec()
+        ) / correction_factor
+        return corrected_current
```

Comment:
Should: Parallelise I/O

```suggestion
        current_gain, voltage_per_sec = await asyncio.gather(
            self.current_amp.get_gain(), self.counter.get_voltage_per_sec()
        )
        correction_factor = self.current_amp.gain_convertion_table[current_gain].value
        corrected_current = voltage_per_sec / correction_factor
        return corrected_current
```